### ISSUE TITLE
[Store gateway] Highly increased latency on "get_range" bucket operations on 0.30.2

### ISSUE THREAD

**Issue by thomas-maurice** [thanos-io/thanos#6540]
**Thanos, Prometheus and Golang version used**:

* Thanos 0.30.2
* Prometheus 2.44.0

**Object Storage Provider**:

Amazon S3

**What happened**:

We recently upgraded Thanos from 0.28.1 to 0.30.2 (we didn't upgrade to the newer 0.31.X version because of the deduplication bug https://github.com/thanos-io/thanos/issues/6257 ). After the upgrade we have seen a dramatic increase in the latency for `get_range` operations on the store gateway. We didn't spot it initially because the environments we tested it in were very low data intensive, however after upgrading our bigger clusters we noticed an increase of the p99 from about a few hundreds milliseconds to sometimes over 2 or even 5 seconds depending on the environment as shown on the graph below

![2023-07-20_10-09](https://github.com/thanos-io/thanos/assets/3941401/13cf9e8a-5a69-4488-8972-0bee226584ee)

Our storegateway pods are configured as follows:

```
    Args:
      store
      --log.level=info
      --log.format=logfmt
      --grpc-address=0.0.0.0:10901
      --http-address=0.0.0.0:10902
      --data-dir=/data
      --objstore.config-file=/conf/objstore.yml
      --max-time=-7d
      --min-time=-1y
```

**What you expected to happen**:

No significant change in the global bucket operations latencies

**How to reproduce it (as minimally and precisely as possible)**:

Not sure how to reproduce, in our case it was a simple matter as upgrading the running container from 0.28.1 to 0.30.2. But we noticed that this behaviour is appearing on our two busiest clusters

**Full logs to relevant components**:

**Anything else we need to know**:

We are running on Amazon EKS

**Comment by fpetkovski** [comment]
Do you have any way to check request latencies in AWS S3 directly? I wonder if the way they are measured has changed between versions. Also has the number of requests gone up after the upgrade?

**Comment by thomas-maurice** [comment]
I couldn't fine any AWS side latency metrics for AWS, we seem to be observing this only on 2 of our 6 environments, and these are the environments that are the most loaded in terms of metrics/queries. I don't think this has to do with a change in how measurements are done because I've seen impact on systems like grafana after the upgrade

**Comment by kiyanabah** [comment]
After upgrading to `Thanos 0.30.2`, we have noticed an incresase in `sent chunk size`,  I was wondering what might be related to increase latency on `get_renge `bucket operation which impact on the operation's performance.  
![image](https://github.com/thanos-io/thanos/assets/18153275/5e83ff9c-36a0-4451-b247-6471dd65a193)

**Comment by mjimeneznet** [comment]
We've downgrade again Thanos to `0.28.1` and eveything went back to previous values. As example:

Request duration between Query and StoreGateway
![image](https://github.com/thanos-io/thanos/assets/5918120/5faaa032-4eb5-4fa5-9ad6-791e940967e8)

Duration on operations in buckets in StoreGateway:
![image](https://github.com/thanos-io/thanos/assets/5918120/645e4e05-c5c4-44b2-9d82-006c548cc30b)

Merge Duration in StoreGateway: 
![image](https://github.com/thanos-io/thanos/assets/5918120/b4822622-1f95-4c57-9b51-ad605bca0987)

Memory usage in StoreGateway:
![image](https://github.com/thanos-io/thanos/assets/5918120/7497728b-18e8-4a93-b7b0-dc48ee1b2362)

CPU usage in Thanos Query:
![image](https://github.com/thanos-io/thanos/assets/5918120/7078cc91-2c01-41b6-a8e1-9803cbb80a75)

Memory usage in Thanos Query:
![image](https://github.com/thanos-io/thanos/assets/5918120/9713caa5-8198-4e6c-98e3-f3b08322fccc)

Chunk Size:
![image](https://github.com/thanos-io/thanos/assets/5918120/77a82e75-5589-4741-beb6-3358b7dee901)

**Comment by douglascamata** [comment]
@thomas-maurice could you try 0.29.1 so that we can look at metrics and try to narrow down the exact release that change this behavior?

**Comment by douglascamata** [comment]
@mjimeneznet would you be kind enough to try 0.30.1, please?

**Comment by kiyanabah** [comment]
> @mjimeneznet would you be kind enough to try 0.30.1, please?

Hi @douglascamata , We've performed upgrades across one environment, moving the Thanos image within both the Prometheus operator and Thanos components to version from` 0.29.0 `to  `0.30.1`. Additionally, we've updated the helm chart to version `11.6.8.`

Following these upgrades, we've observed a noticeable increase in **chunk size** and occurrences of the **get_range** bucket operation. As a result, this increase has triggered the **ThanosStoreObjstoreOperationLatencyHigh** [alert](https://github.com/thanos-io/thanos/blob/main/examples/alerts/alerts.yaml#L374).
![memory-usage](https://github.com/thanos-io/thanos/assets/18153275/bab497e4-3e24-4987-a541-d8e4a32d214d)
![cpu-usage](https://github.com/thanos-io/thanos/assets/18153275/ac33db34-ea43-4942-838b-2b97307da03e)
![store](https://github.com/thanos-io/thanos/assets/18153275/9d513d68-63d5-4f63-bfb7-42f7051d438a)

**Comment by kiyanabah** [comment]
> Kind reminder: **do not** run 0.30.1 for long on a production environment. You want 0.30.2 there asap to have a very important fix from #6086. Thanks a lot for testing it though.

I appreciate your attention to the issue. We did downgrade to 0.29.0 as we have it more stable for now. Thank you :blush:

**Comment by fpetkovski** [comment]
@thomas-maurice could you post the contents of the `/conf/objstore.yml` file?

**Comment by thomas-maurice** [comment]
@fpetkovski it is pretty stock, we didn't customise it

```yaml
---
type: s3
config:
  bucket: [BUCKET NAME]
  endpoint: [S3 ENDPOINT]
  aws_sdk_auth: true
```

And that's it, so I am assuming the storegateway is working with the default config values

**Comment by thomas-maurice** [comment]
@fpetkovski any ideas of things we could try to troubleshoot further more ?

**Comment by fpetkovski** [comment]
The only thing that comes to mind is to try out 0.32 once we release it. We are waiting on https://github.com/thanos-io/thanos/pull/6317 before we can cut a new release, but there have been many changes since 0.30.2 and the issue might have been addressed already.

**Comment by thomas-maurice** [comment]
Okay :) Will do !

**Comment by lasermoth** [comment]
@thomas-maurice with `v0.32.0` / `v0.32.2`, I was wondering if you had a change to test this yet?

**Comment by thomas-maurice** [comment]
@lasermoth not yet, I'll update the issue when we had time to try it out !

**Comment by neitrinoweb** [comment]
Hello! Any information if this has been fixed in 32.5?

**Comment by MichaHoffmann** [comment]
> Hello! Any information if this has been fixed in 32.5?

There were some improvements but the general method to iterate the bucket has not changed. There is the idea of defaulting to previous behaviour and enabling current one ( for cases where it helps ) with a hidden flag and for long term add a bucket index so we can sync cheaply but no work has been done yet.

**Comment by thomas-maurice** [comment]
Hello ! Sorry for the last reply but yes this was fixed in subsequent Thanos versions!

I'm closing this
